tidy3d/plugins/mode/mode_solver.py

```diff
@@ -317,27 +318,410 @@ def data_raw(self) -> ModeSolverData:
+                    fields["Hr"] * cos_theta - fields["Htheta"] * sin_theta
+                )
+                rotated_fields[h_field_keys[1]][idx_x, idx_y, idx_z, f_idx, mode_idx] = phase * (
+                    fields["Hr"] * sin_theta + fields["Htheta"] * cos_theta
+                )
+                rotated_fields[h_field_keys[2]][idx_x, idx_y, idx_z, f_idx, mode_idx] = (
+                    phase * fields["Haxial"]
+                )
+
+    def transform_field_components(
+        self, mode_base, mode_target, n_complex, bend_center, radius, theta_ref, bend_axis
+    ):
+        x, y, z, f, mode_index = (
+            mode_target.Ex[coord] for coord in ["x", "y", "z", "f", "mode_index"]
+        )
+
+        # Initialize output arrays
+        shape = (x.size, y.size, z.size, f.size, mode_index.size)
+        rotated_fields = {
+            name: np.zeros(shape, dtype=complex) for name in ["Ex", "Ey", "Ez", "Hx", "Hy", "Hz"]
+        }
+
+        # Mapping for each combination of bend_axis and inj_axis
+        calculation_map = {
+            (2, 1): (x.values - bend_center[0], y.values[0] - bend_center[1]),
+            (2, 0): (x.values[0] - bend_center[0], y.values - bend_center[1]),
+            (0, 1): (y.values[0] - bend_center[1], z.values - bend_center[2]),
+            (0, 2): (y.values[0] - bend_center[1], z.values - bend_center[2]),
+            (1, 0): (x.values[0] - bend_center[0], z.values - bend_center[2]),
+            (1, 2): (x.values - bend_center[0], z.values[0] - bend_center[2]),
+        }
+
+        if (bend_axis, self.normal_axis) in calculation_map:
+            self.calculate_fields(
+                rotated_fields,
+                mode_base,
+                n_complex,
+                mode_index,
+                f,
+                theta_ref,
+                radius,
+                *calculation_map[(bend_axis, self.normal_axis)],
+                self.normal_axis,
+                bend_axis,
+            )
+
+        # Convert numpy arrays to xarray DataArrays
+        coords = {"x": x, "y": y, "z": z, "f": f, "mode_index": mode_index}
+        rotated_data_arrays = {
+            name: xr.DataArray(
+                rotated_fields[name], coords=coords, dims=("x", "y", "z", "f", "mode_index")
+            )
+            for name in rotated_fields
+        }
+
+        return mode_target.updated_copy(**rotated_data_arrays)
+
+    def theta_reference(bend_axis, bend_center, monitor_center):
+        axis_map = {0: (2, 1), 1: (2, 0), 2: (1, 0)}
+        coord1, coord2 = axis_map[bend_axis]
+        theta_ref = np.arctan2(
+            monitor_center[coord1] - bend_center[coord1],
+            monitor_center[coord2] - bend_center[coord2],
+        )
+        return theta_ref
+
+    def compute_bent_center(self):
+        bend_axis = self.mode_spec.bend_axis
+        id_inj, id_mnt_plane = self.pop_axis((0, 1, 2), axis=self.size.index(0.0))
+        id_bent = id_mnt_plane[bend_axis]
```

Comment:
I think you call `id_bent` here `idx_bend` above in the `data_raw` function, and then you call it `id_bend` in the `rotate_structure` function. Apart from keeping things consistent, it may be better to just gather this in a small property of the ModeSolver, similar to `self.normal_axis`. Note that `id_mnt_plane` are unused anywhere else in these functions, while `id_inj` is just `self.normal_axis`, so a property that just returns the `id_bend` (maybe call it `bend_axis_3d` or something like that) should suffice.